main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -34,20 +40,350 @@ import ca.uwaterloo.flix.util.Validation._
+      case Binary(_, _, _, _, _) => None
+      case IfThenElse(_, _, _, _, _) => None
+      case Let(_, _, _, _, _) => None
+      case LetRec(_, _, _, _, _) => None
+      case Is(_, _, _, _) => None
+      case Tag(_, _, _, _, _) => None
+      case Untag(_, _, _, _, _) => None
+      case Index(_, _, _, _) => None
+      case Tuple(_, _, _) => None
+      case Existential(_, _, _) => None
+      case Universal(_, _, _) => None
+      case NativeConstructor(_, _, _, _) => None
+      case NativeField(_, _, _) => None
+      case NativeMethod(_, _, _, _) => None
+      case UserError(_, _) => None
+      case MatchError(_, _) => None
+      case SwitchError(_, _) => None
+      /* Error */
+      case MkClosure(_, _, _, _) =>
+        throw InternalCompilerException(s"Unexpected expression $exp after lambda lifting")
+    }
+  }
+
+  /**
+    * Traverses through `exp` and performs inlining using
+    * definitions and sizes from `scores`
+    */
+  def inlineExpr(definitions: Map[DefnSym, Constant], scores: Map[DefnSym, Int], exp: Expression)(implicit genSym: GenSym): Expression =
+  {
+    def f = inlineVisitor(definitions,scores,_: Expression)
+    AstVisitor.visitExpressionWith(f, exp)
   }
 
+}
+
+object FreshenExpr {
+
+  /**
+    * Freshen all let binds in an expression
+    */
+  def freshenExpr(exp: Expression)(implicit genSym: GenSym): Expression =
+  {
+    val (m,newLets) = freshSymbols(exp)
+    Inliner.substituteWith(m,newLets)
+  }
+
+  def freshSymbols(exp:Expression)(implicit genSym: GenSym): (Map[VarSym, VarSym], Expression) =
+  {
+    var m = Map.empty[VarSym, VarSym]
+
+    def freshExprVisitor(exp3:Expression): Option[Expression] =
+    {
+      exp3 match {
+        case Let(sym, exp1, exp2, tpe, loc) =>
+          val newSym = Symbol.freshVarSym("inlnrFresh")
+          m += (sym -> newSym)
+          val freshExp1 = AstVisitor.visitExpressionWith(freshExprVisitor, exp1)
+          val freshExp2 = AstVisitor.visitExpressionWith(freshExprVisitor, exp2)
+          Some(Let(newSym,freshExp1,freshExp2,tpe,loc))
+        case LetRec(sym, exp1, exp2, tpe, loc) =>
+          val newSym = Symbol.freshVarSym("inlnrFresh")
+          m += (sym -> newSym)
+          val freshExp1 = AstVisitor.visitExpressionWith(freshExprVisitor, exp1)
+          val freshExp2 = AstVisitor.visitExpressionWith(freshExprVisitor, exp2)
+          Some(LetRec(newSym,freshExp1,freshExp2,tpe,loc))
+        case _ => None
+      }
+    }
+
+    val newLets = AstVisitor.visitExpressionWith(freshExprVisitor, exp)
+    (m,newLets)
+  }
+
+}
+
+object AstVisitor {
+
+  /**
+    * Expression Visitor function for SimplifiedAST
+    *
+    * @param f function that takes in a Expression and returns a Option[Expression]
+    * @param exp Expression to visit
+    * @return if f(exp) is Some(e) then e, else recursively visit expressions in exp
+    */
+  def visitExpressionWith(f: Expression => Option[Expression], exp: Expression): Expression =
```

Comment:
Ok; this is the biggest issue with the code as I see it now. I have thought long about how to come up with a suitable visitor abstraction, but it always breaks down. The essentially problem seems to be the following:

1. I *always* want pattern matches to be exhaustive without the use of wildcards. We absolutely *cannot* use wildscards because it means that when the AST changes (e.g. a new type of node is added) then we don't get any warnings (because the new node is matched by the wildcard) and it becomes very difficult to track down all the places where the new node must be handled.
2. This means that we always have to write out a full pattern match. In that case, we might always well have full control over the recursion ourselves. This also makes it easier to control exactly how the recursion operations and what arguments are passed in and out. 
3. If every function is a structural recursion on the ast then the compiler is very easy to understand and follow. Yes, there is more typing, but you can understand a function by looking at in isolation.